File: squaregaussian_scripts.m
%% Scripts to run ode45 for ray tracing
% This is just for the square case and the metric g is a Gaussian. 
% First define the function:

% function dH = gaussianmetric(s,u)
% % u(1) = x coordinate, u(2) = y coordinate, 
% % u(3) = x velocity, u(4) = y velocity.
%     dH = zeros(4,1);
%     % The positions:
%     dH(1) = u(3).*exp(u(1).^2 + u(2).^2);
%     dH(2) = u(4).*exp(u(1).^2 + u(2).^2);
%     % The momenta: 
%     dH(3) = (u(3).^2 + u(4).^2).*u(1).*exp(u(1).^2 + u(2).^2);
%     dH(4) = (u(3).^2 + u(4).^2).*u(2).*exp(u(1).^2 + u(2).^2);
% end

%% Checking the ODE evolution is fine and iterating.
tic;
Nedge = 2^5;   % number of partitions of sides.
Nangle = 2^5;   % number of partitions of angle.
dphi = pi/Nangle;   % This needs to be taken better care of
dl = 2/Nedge; 
% Should double for loop over n,N!
% Can switch the order of for loops to see the change in ray traces with 
% entry position or entry angle graphically.
%%
% To terminate when either |x|,|y| >= 1
options = odeset('Events',@sgEventsFcn); 
%Here,
% function [value,isterminal,direction] = sgEventsFcn(s,u)
% % Terminate ODE evolution when the position is outside of square.
% value = max(0,1 - max(abs(u(1)), abs(u(2))));
% isterminal = 1;
% direction = -1;
% end
%%
for i = 1:Nedge-2
    clf
    for j = 1:Nangle-2
        % u0 = [-1;1-i*dl;cos(j*dphi-pi/2); sin(j*dphi-pi/2)];   % Only for left edge!
        % u0 = [-1+i*dl;-1;cos(j*dphi); sin(j*dphi)];   % This is for bottom edge!
         u0 = [1;-1+i*dl;cos(j*dphi+pi/2); sin(j*dphi+pi/2)];   % Only for right edge!
        % u0 = [1-i*dl;1;cos(-j*dphi); sin(-j*dphi)];   % This is for top edge!
        [s,u,~,ue] = ode45(@gaussianmetric, [0,1], u0, options);
        % Noting how some rays don't make it out yet
        plot(u(:,1), u(:,2)); axis([-2 2 -2 2]);
        xlabel('x position'); ylabel('y position');
        x = -1:0.1:1;
        y = ones(length(x),1);
        hold on
        plot(x,y); plot(x,-y); plot(y,x); plot(-y,x);
        pause(0.01);
    end
end
toc;
% The last spread of angles is displayed

%% Scattering Relation, Exit Data. 
% Then search for when it FIRST leaves the box somehow. 
% Uses squaregaussianrelation.m --> if working correctly, then use to
% get all the data with SGscatteringrelation.m 

% Two velocity interpolations used in squaregaussianrelation.m , 
% Could subtract two methods to see how their accuracy is w.r.t. each other

tic;
ds = 1;  % Or make it smaller if desired?
uNSEW = cell((Nedge-2),(Nangle-2)); 
% The cell of all exit-pos/vel for each given incidence pos/angle. 
% To access: Use for ex. cellfun(@(v) v(4), uNSEW(:,1)) ?
% For the cells, each row is a point on the boundary and each collumn is
% the angle of incidence. So,
% The 1st entry ~ location on the edge. 2nd entry ~ angle of incidence.

for i = 1:Nedge-2
    for j = 1:Nangle-2
        % u0 = [-1; 1-i*dl; cos(j*dphi - pi/2); sin(j*dphi - pi/2)];   % Only for left (West) edge!
        % u0 = [-1+i*dl;-1;cos(j*dphi); sin(j*dphi)];   % This is for bottom (South) edge!
         u0 = [1;-1+i*dl;cos(j*dphi+pi/2); sin(j*dphi+pi/2)];   % Only for right (East) edge!
        % u0 = [1-i*dl;1;cos(-j*dphi); sin(-j*dphi)];   % This is for top (North) edge!
        uNSEW{i,j} = squaregaussianrelation(u0,ds);
    end
end
toc;
%%
% To check qualitatively with the plot above:
celldisp(uNSEW(end,:))
% Be careful if we wanted to switch the order of for loops
% It looked consistent with all 4 sides. 
%% Entire scattering relation
% Use: (Takes a while to compute)
tic;
[uW, uN, uE, uS] = SGscatteringrelation(Nedge, Nangle);
toc;
% To check with prior notes I guess:
celldisp(uE(end,:))

File: SquareGaussian/RayTracing/LegacyODE/squaregaussian_scripts.jl
using ODE
using PyPlot
include("RayTracing.jl");

Nedge = 2^4;   # number of partitions of sides.
Nangle = 2^5;   # number of partitions of angle.
dphi = pi/Nangle;   # This needs to be taken better care of
dl = 2/Nedge;
# options = odeset("Events",sgEventsFcn);


### Trying Events in ODE.jl:
u0 = [-1,0,1/2^0.5,1/2^0.5];
s0 = 0.0
~,u = ODE.ode45(gaussianmetric,u0,[0.,2.],stopevent = (s,u) -> (abs(u[1]) > 1 || abs(u[2]) > 1));
u1 = map(z->z[1],u);
u2 = map(z->z[2],u);
plot(u1,u2);
ax = gca();
  ax[:set_xlim]([-2,2]);
  ax[:set_ylim]([-2,2]);
  ax[:set_xlabel]("x-position");
  ax[:set_ylabel]("y-position");
x = -1:0.1:1;
y = ones(length(x),1);
plot(x,y); plot(x,-y); plot(y,x); plot(-y,x);
# Event seems to not trigger root finding :(

# Another option is to try using
# isoutofdomain(y) = abs(y[1]) > 1 || abs(y[2]) > 1;  # adding isnan(y) makes it a vector...

# Another approach is to define each approach to solving:
ode=ODE.ExplicitODE(s0,u0,gaussianmetric)
stepper=ODE.RKStepperAdaptive{:rk45,Float64}()
opts=ODE.Options{Float64}(tstop=2.0)

for (s,u) in ODE.solve(ode,stepper,opts)    # Currently, not recognizing RHS of gaussianmetric?
  if abs(u[1]) > 1 || abs(u[2]) > 1
    print((s,u))
    break
  end
end

## This is not working yet ... ##

u1 = map(z->z[1],u);
u2 = map(z->z[2],u);
plot(u1,u2);
ax = gca();
  ax[:set_xlim]([-2,2]);
  ax[:set_ylim]([-2,2]);
  ax[:set_xlabel]("x-position");
  ax[:set_ylabel]("y-position");
x = -1:0.1:1;
y = ones(length(x),1);
plot(x,y); plot(x,-y); plot(y,x); plot(-y,x);


### PART 1: Just for checking the ODE is setup and evolving correctly

for i = 1:Nedge-1
    clf()
    for j = 1:Nangle-1
        # u0 = [-1;1-i*dl;cos(j*dphi-pi/2); sin(j*dphi-pi/2)];   # Only for left edge!
         u0 = [-1+i*dl;-1;cos(j*dphi); sin(j*dphi)];   # This is for bottom edge!
        # u0 = [1;-1+i*dl;cos(j*dphi+pi/2); sin(j*dphi+pi/2)];   # Only for right edge!
        # u0 = [1-i*dl;1;cos(-j*dphi); sin(-j*dphi)];   # This is for top edge!

        # ~,u = ode45(gaussianmetric, u0, [0.0,1.0]);  # For official ODE.jl
        ~,u = ODE.ode45(gaussianmetric, u0, [0.0,1.0], stopevent = (s,u) -> (abs(u[1]) > 1 || abs(u[2]) > 1));

        # Noting how some rays don't make it out yet
        u1 = map(z->z[1],u);
        u2 = map(z->z[2],u);
        plot(u1,u2);
        ax = gca();
        if j == 1
          ax[:set_xlim]([-2,2]);
          ax[:set_ylim]([-2,2]);
          ax[:set_xlabel]("x-position");
          ax[:set_ylabel]("y-position");
        x = -1:0.1:1;
        y = ones(length(x),1);
        plot(x,y); plot(x,-y); plot(y,x); plot(-y,x);
        end
        pause(0.0000001);
    end
end

### PART 2:  This is for just extracting the exit data for each entrance data

ds = 1;  # Or can make it smaller/larger?
uNSEW = cell((Nedge-1),(Nangle-1));
for i = 1:Nedge-1
    for j = 1:Nangle-1
        # u0 = [-1; 1-i*dl; cos(j*dphi - pi/2); sin(j*dphi - pi/2)];   # Only for left (West) edge!
         u0 = [-1+i*dl;-1;cos(j*dphi); sin(j*dphi)];   # This is for bottom (South) edge!
        # u0 = [1;-1+i*dl;cos(j*dphi+pi/2); sin(j*dphi+pi/2)];   # Only for right (East) edge!
        # u0 = [1-i*dl;1;cos(-j*dphi); sin(-j*dphi)];   # This is for top (North) edge!
        uNSEW[i,j] = squaregaussianrelation(u0,ds);
    end
end

# Check with the last graphic:
print(uNSEW[end,:])

# To access the individual components:
# xexit = map(x -> x[1], uNSEW);  # Or also xexit = [x[1] for x in uNSEW];
# yexit = map(x -> x[2], uNSEW);
# vxexit = map(x -> x[3], uNSEW);
# vyexit = map(x -> x[4], uNSEW);

### Part 3: Entire scattering relation
# Use the following:
uW, uS, uE, uN = SGscatteringrelation(Nedge,Nangle,ds);

# Check with the prior one (need to pick NSEW)
print(uS[end,:] - uNSEW[end,:])

#### Now ... how to imitate ODE's events in Matlab in Julia ... use norm??
# Like, the norm is dist(point, boundary) and if it is within a small tolerance
# it is considered at / out of the boundary.
